### PR TITLE
Prepare DID storage for multiple DIDs per DID Subject

### LINKED ISSUE #3193: Add `subject_id` to `did` table. Add `findBySubjectId`
Same subject DIDs should automatically add AlsoKnownAs of other DIDs to documents.

### PR DESCRIPTION
**Pull request by woutslakhorst**
closes #3193  

- removed selfControl & controllers for did:nuts
- removed tenant option for did:web
- moved sql storage from did:web to vdr

### CODE REVIEW

File: vdr/didweb/manager.go [hunk 1]
```diff
@@ -107,64 +113,62 @@ func (m Manager) AddVerificationMethod(_ context.Context, _ did.DID, _ managemen
...
+	vmAsJson, err := json.Marshal(verificationMethod)
 	if err != nil {
 		return nil, nil, err
 	}
-	if err := m.store.create(newDID, *verificationMethod); err != nil {
-		return nil, nil, fmt.Errorf("store new DID: %w", err)
+
+	sqlDid := sql.DID{
+		ID:      newDID.String(),
+		Subject: newDID.String(),
```
Comment (reinkrul) [inline vdr/didweb/manager.go:156]: shouldn't this be a UUID (subject) and be supplied through the options?

File: vdr/didweb/manager.go [hunk 2]
```diff
@@ -107,64 +113,62 @@ func (m Manager) AddVerificationMethod(_ context.Context, _ did.DID, _ managemen
...
+	}
+	var doc *sql.DIDDocument
+	if doc, err = documentStore.CreateOrUpdate(sqlDid, []sql.SqlVerificationMethod{{
```
Comment (reinkrul) [inline vdr/didweb/manager.go:159]: You're not really creating a new DID document here? You could have multiple DIDs (even multiple did:web's?) belonging to the same subject, which should then all return the same services and VMs?

File: vdr/didweb/manager.go [hunk 3]
```diff
@@ -107,64 +113,62 @@ func (m Manager) AddVerificationMethod(_ context.Context, _ did.DID, _ managemen
...
 	}
-	parsedNewDID, err := did.ParseDID(newDID.String()) // make sure internal state is consistent (DecodedID)
 	if err != nil {
-		return nil, nil, fmt.Errorf("invalid new DID: %s: %w", newDID.String(), err)
+		return nil, nil, fmt.Errorf("parse new DID: %w", err)
 	}
-	return m.create(ctx, *parsedNewDID)
-}
+	tx := m.db.Begin()
+	defer tx.Rollback()
```
Comment (reinkrul) [inline vdr/didweb/manager.go:133]: better use `db.Transaction(func())`

File: vdr/sql/did.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	// DeleteAll removes all DIDs for a subject from the database
+	DeleteAll(subject string) error
+	// Find returns a DID by its ID
+	Find(id did.DID) (*DID, error)
+	// FindBySubject returns all DIDs for a subject
+	FindBySubject(subject string) ([]DID, error)
+}
+
+// SqlDIDManager is the implementation of the DIDManager interface
+type SqlDIDManager struct {
```
Comment (reinkrul) [inline vdr/sql/did.go:59]: already in `sql` package, so no need to prepend types with `Sql`

**Review by reinkrul — CHANGES_REQUESTED** [review]
How are we going to keep services and VMs over multiple DIDs belonging to the same subject in sync?

**Comment by woutslakhorst** [comment]
> How are we going to keep services and VMs over multiple DIDs belonging to the same subject in sync?

Use the DB as primary and (for now) the network update as side-effect (maybe within the sql transaction)

File: vdr/didweb/manager.go [hunk 5]
```diff
@@ -30,39 +31,28 @@ import (
...
 }
 
 var _ management.DocumentManager = (*Manager)(nil)
 
 // NewManager creates a new Manager to create and update did:web DID documents.
 func NewManager(rootDID did.DID, tenantPath string, keyStore crypto.KeyStore, db *gorm.DB) *Manager {
 	return &Manager{
-		store:      &sqlStore{db: db},
+		db: db,
+		//store:      &sqlStore{db: db},
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/didweb/manager.go:55]: leftover, remove

File: vdr/didweb/manager.go [hunk 6]
```diff
@@ -71,23 +61,39 @@ func NewManager(rootDID did.DID, tenantPath string, keyStore crypto.KeyStore, db
...
+	documentStore := sql.NewDIDDocumentManager(tx)
+	sqlDocument, err := documentStore.Latest(subjectDID)
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return resolver.ErrNotFound
+		}
+		return err
+	}
+	if sqlDocument == nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/didweb/manager.go:84]: can't, remove

File: vdr/didweb/manager.go (hunk 1 above)
Comment (woutslakhorst) [reply vdr/didweb/manager.go:156]: correct, but allowing a subject to be passed through the API would be part of another PR. Added a todo as comment

File: vdr/didweb/manager.go (hunk 2 above)
Comment (woutslakhorst) [reply vdr/didweb/manager.go:159]: True, but I wanted this PR to just update the DB model. In a later PR we'll move this logic one level higher (so the web manager will no longer have a create just a read/list/resolve)

File: vdr/sql/did.go (hunk 4 above)
Comment (woutslakhorst) [reply vdr/sql/did.go:59]: true, but `DIDManager` is the interface. When declaring vars from other packages you should use the interface. We won't see `var a sql.SqlDIDManager` anywhere. Also I'm not sure the package will not be renamed after some refactors.

File: vdr/didweb/manager.go [hunk 7]
```diff
@@ -71,23 +60,36 @@ func NewManager(rootDID did.DID, tenantPath string, keyStore crypto.KeyStore, db
...
+	db         *gorm.DB
 	rootDID    did.DID
-	store      store
 	keyStore   crypto.KeyStore
 	tenantPath string
 }
 
 func (m Manager) Deactivate(ctx context.Context, subjectDID did.DID) error {
-	verificationMethods, _, err := m.store.get(subjectDID)
+	tx := m.db.Begin()
```
Comment (reinkrul) [inline vdr/didweb/manager.go:70]: db.Transaction(func())

**Review by reinkrul — APPROVED** [review]

**Pull request merged**